pkg/cluster/kubernetes/resource/load.go

```diff
@@ -127,11 +130,18 @@ func looksLikeChart(dir string) bool {
 
 // ParseMultidoc takes a dump of config (a multidoc YAML) and
 // constructs an object set from the resources represented therein.
-func ParseMultidoc(multidoc []byte, source string) (map[string]KubeManifest, error) {
+func ParseMultidoc(multidoc []byte, source string, sopsEnabled bool) (map[string]KubeManifest, error) {
 	objs := map[string]KubeManifest{}
-	decoder := yaml.NewDecoder(bytes.NewReader(multidoc))
 	var obj KubeManifest
 	var err error
+	cleartext := multidoc
+	if sopsEnabled {
+		cleartext, err = softSopsDecrypt(multidoc)
```

Comment:
Shouldn't we also call `softSopsDecrypt()` after https://github.com/fluxcd/flux/pull/2580/files#diff-c3f2eeebdc2b683408b96d1ed03621c5R156 in case we are decoding a document stream which isn't which contains individually encrypted documents?

In fact, how does sops encryption of document streams work? If it's one to one (i.e. when encrypting a stream of n documents leads to another stream of n encrypted documents) , then we should only call it there, since the stream itself is parseable in a normal fashion.

Reply:
According to the sops [documentation](https://github.com/mozilla/sops#33yaml-streams), the hash isn't computed for each individual document. It encrypts the entire file, and saves the hash of that. So tampering with one document would cause all of them to fail the validation step in the decryption.

In order to decrypt each individual document, we would have to unmarshal the yaml stream, and then marshal every document in order to try the softSopsDecrypt. This approach would work if a bunch of individually encrypted yaml files were joined in a stream and passed to the function, but in this case yaml documents that were encrypted in a yaml stream would not pass the hash validation step.

One interesting thing to note is that sops actually includes the same metadata in all the encrypted documents in a stream. So it would technically be possible to match the documents, but it wouldn't be pretty.

Reply:
I have been playing with `sops` and it indeed fails to decrypt isolated documents from an encrypted yaml stream, with a `MAC mismatch` error.

Also, `sops` doesn't transparently decrypt streams composed by encrypted and plain yaml manifests.

> So it would technically be possible to match the documents, but it wouldn't be pretty.

Yes, I think that's trying to be too smart.

With these restrictions, I can't think of a good way to make sops work well with manifest generation (at least with the current manifest generation scheme; in which generators print yaml streams to `stdout`).

I see two options:

1. Explicitly avoid automatic decryption for generated files. We decrypt sops files automatically for raw manifests (i.e. only in the Loadbut we also include the sops binary, to be used in `.flux.yaml` files.
2. Use the in-place decrypt-first scheme. However, I still think that trying to decrypt every file is overkill. Having a specific file suffix (like the PR originally proposed) may solve the problem.

I think we should go for (1), having raw sops manifest files being decrypted already adds a lot of value. Manifest generation can use the binary.

I would like to know @squaremo 's opinion on this one though

Reply:
BTW, @PaulFarver thanks a lot for enduring so many comments and design changes 🌷

Reply:
@2opremio No worries. I think it is awesome to see this level of interest 🥇

Reply:
I think it would be perfectly fine to support:

 1. SOPS decryption of plain YAMLs;
 2. calling `sops` as a command before running `kustomize build .`

I don't see a strong use case for decrypting the output of arbitrary commands. The conclusion I reach from the arguments in the thread above is that it's just too fiddly to organise for a correctly encrypted set of bytes, and you are much better off just doing 1) or 2).

I think this indicates that sops decryption should be done when reading a file in `cluster/kubernetes/resource.Load` (since that loads from files, and SOPS deals with files), but not attempted in ParseManifests.

If it can be written as a wrapper around `Manifests`, that might be a little neater than passing a boolean down through a lot of procedures.

Reply:
OK, it seems @squaremo and I agree (which doesn't always happen, so it's a good sign :) )

@PaulFarver can you please:
1) Move sops decryption back to `Load()`
2) Document (in the `--sops` flag help) that automatic `sops` decryption is not applied when using `.flux.yaml` files , and that the `sops` binary can be used for that.
3) Include the `sops` binary in the Flux image
4) Add test(s)

Reply:
> 3\. Include the `sops` binary in the Flux image

Why this?

Reply:
Because otherwise you cannot invoke `sops` from the `.flux.yaml` files without building your own image (or running an init container).

Reply:
Oh OK. I am a bit reluctant to keep adding binaries, but it is suggested by "support [...] calling `sops` as a command before running `kustomize build .`", so fair cop :-S